libs/mpi/CMakeLists.txt

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2019 The STE||AR-Group
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+cmake_minimum_required(VERSION 3.3.2 FATAL_ERROR)
+
+# if networking is off, or MPI parcelport is not being used
+# then allow the option of using an MPI
+if ("${HPX_WITH_NETWORKING}" AND "${HPX_WITH_PARCELPORT_MPI}")
+  hpx_option(HPX_WITH_MPI_FUTURES BOOL
+    "Enable support for returning futures from MPI asynchronous calls"
+    " (default: ON if networking+MPI already enabled)"
+    ON CATEGORY "MPI")
+else()
+  hpx_option(HPX_WITH_MPI_FUTURES BOOL
+    "Enable support for returning futures from MPI asynchronous calls"
+    " (default: ON if networking+MPI already enabled)"
+    OFF CATEGORY "MPI")
+endif()
+
+# if the user does not want mpi async futures, quit - the module will not be enabled
+if (NOT ${HPX_WITH_MPI_FUTURES})
+  return()
+endif()
+
+#
+# only process further when enabled
+#
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+
+find_package(MPI REQUIRED)
+
+if(${CMAKE_VERSION} VERSION_LESS "3.12.0")
+    message(ERROR "cmake 3.12 or newer required")
+endif()
+
+# All cmake version don't have the same found variable set
+if(NOT MPI_FOUND AND NOT MPI_C_FOUND)
+  hpx_error("MPI could not be found and hpx::mpi library was requested\n
+  set MPI_MPI_CXX_COMPILER to point to your MPI installation's mpicxx")
+endif()
+
+# if networking is on, then the mpi target will already be setup
+# so only do it if it doesn't exist
+if (NOT TARGET hpx::mpi )
+  add_library(hpx::mpi INTERFACE IMPORTED)
+  set_property(TARGET hpx::mpi PROPERTY
+    INTERFACE_INCLUDE_DIRECTORIES ${MPI_INCLUDE_PATH} ${MPI_C_INCLUDE_DIRS})
+
+  if (MPI_C_LIBRARIES)
+      target_link_libraries(hpx::mpi INTERFACE ${MPI_CXX_LIBRARY})
+  endif()
+  # Ensure compatibility with older versions
+  if (MPI_LIBRARY)
+    target_link_libraries(hpx::mpi INTERFACE ${MPI_LIBRARY})
+  endif()
+  if (MPI_EXTRA_LIBRARY)
+    target_link_libraries(hpx::mpi INTERFACE ${MPI_EXTRA_LIBRARY})
+  endif()
+
+  set_property(TARGET hpx::mpi PROPERTY INTERFACE_COMPILE_OPTIONS ${MPI_CXX_COMPILE_FLAGS})
+    set_property(TARGET hpx::mpi PROPERTY INTERFACE_COMPILE_DEFINITIONS ${MPI_CXX_COMPILE_DEFINITIONS})
+endif()
+
+# Default location is $HPX_ROOT/libs/mpi/include
+set(mpi_headers
+    hpx/mpi/mpi_executor.hpp
+    hpx/mpi/mpi_future.hpp
+)
+
+# Default location is $HPX_ROOT/libs/mpi/include_compatibility
+set(mpi_compat_headers)
+
+set(mpi_sources)
+
+include(HPX_AddModule)
+add_hpx_module(mpi
```

Comment:
Was the plan to call this `mpi_support`? Or should we stick with `mpi`?

Reply:
Originally I thought of MPI_Support, but I think mpi is fine, though I would like the CMake variable to be `HPX_WITH_MPI_SUPPORT`. I need to write some docs to explain a few things - like what happens if you enable `HPX_WITH_MPI_SUPPORT` and at the same time have `HPX_WITH_NETWORKING=ON` and MPI PARCELPORT enabled - this has not been tested - it should work, but users would need to be super careful about which communicators they use. I suspect we need to isolate the communicator we use in the `MPI_Testany` and suchlike. Making some notes on this would be sensible. A blog articles with the same info in too (as discussed yesterday).